File: apps/imbib/PublicationManagerCore/Sources/PublicationManagerCore/BibTeX/JournalMacros.swift
//
//  JournalMacros.swift
//  PublicationManagerCore
//
//  Created by Claude on 2026-01-04.
//

import Foundation

/// Expands AASTeX journal macro abbreviations to full journal names.
///
/// Based on the standard AASTeX journal macros from the AAS (American Astronomical Society).
/// See: https://ui.adsabs.harvard.edu/help/actions/journal-macros
public enum JournalMacros {

    // MARK: - Macro Definitions

    /// Dictionary mapping journal macros to full names.
    /// Keys are lowercase without the backslash (e.g., "apj" not "\\apj").
    public static let macros: [String: String] = [
        // Major Astronomical Journals
        "aj": "Astronomical Journal",
        "apj": "Astrophysical Journal",
        "apjl": "Astrophysical Journal, Letters",
        "apjlett": "Astrophysical Journal, Letters",
        "apjs": "Astrophysical Journal, Supplement",
        "apjsupp": "Astrophysical Journal, Supplement",
        "mnras": "Monthly Notices of the Royal Astronomical Society",
        "aap": "Astronomy and Astrophysics",
        "astap": "Astronomy and Astrophysics",
        "aaps": "Astronomy and Astrophysics, Supplement",
        "pasp": "Publications of the Astronomical Society of the Pacific",
        "pasj": "Publications of the Astronomical Society of Japan",
        "pasa": "Publications of the Astronomical Society of Australia",

        // Review Journals
        "araa": "Annual Review of Astronomy and Astrophysics",
        "aapr": "Astronomy and Astrophysics Reviews",

        // Society Publications
        "baas": "Bulletin of the American Astronomical Society",
        "memras": "Memoirs of the Royal Astronomical Society",
        "qjras": "Quarterly Journal of the Royal Astronomical Society",
        "jrasc": "Journal of the Royal Astronomical Society of Canada",
        "memsai": "Memorie della Societa Astronomica Italiana",
        "iaucirc": "IAU Circulars",

        // Specialized Fields
        "icarus": "Icarus",
        "psj": "Planetary Science Journal",
        "solphys": "Solar Physics",
        "jcap": "Journal of Cosmology and Astroparticle Physics",
        "planss": "Planetary and Space Science",
        "grl": "Geophysical Research Letters",
        "jgr": "Journal of Geophysical Research",

        // Physics Journals
        "prl": "Physical Review Letters",
        "pra": "Physical Review A",
        "prb": "Physical Review B",
        "prc": "Physical Review C",
        "prd": "Physical Review D",
        "pre": "Physical Review E",
        "physrep": "Physics Reports",
        "physscr": "Physica Scripta",
        "nphysa": "Nuclear Physics A",

        // Optics and Spectroscopy
        "ao": "Applied Optics",
        "applopt": "Applied Optics",
        "jqsrt": "Journal of Quantitative Spectroscopy and Radiative Transfer",
        "procspie": "Proceedings of the SPIE",

        // General Science
        "nat": "Nature",
        "sci": "Science",
        "ssr": "Space Science Reviews",
        "apss": "Astrophysics and Space Science",

        // Regional Journals
        "actaa": "Acta Astronomica",
        "azh": "Astronomicheskii Zhurnal",
        "sovast": "Soviet Astronomy",
        "caa": "Chinese Astronomy and Astrophysics",
        "cjaa": "Chinese Journal of Astronomy and Astrophysics",
        "rmxaa": "Revista Mexicana de Astronomia y Astrofisica",
        "bac": "Bulletin of the Astronomical Institutes of Czechoslovakia",
        "bain": "Bulletin Astronomical Institute of the Netherlands",
        "zap": "Zeitschrift fuer Astrophysik",

        // Other
        "na": "New Astronomy",
        "nar": "New Astronomy Review",
        "skytel": "Sky & Telescope",
        "aplett": "Astrophysics Letters",
        "apspr": "Astrophysics Space Physics Research",
        "fcp": "Fundamental Cosmic Physics",
        "gca": "Geochimica et Cosmochimica Acta",
        "jcp": "Journal of Chemical Physics",
    ]

    // MARK: - Expansion

    /// Expands a journal field value to its full name.
    ///
    /// This handles multiple formats:
    /// - LaTeX macro: `\apj` → "Astrophysical Journal"
    /// - Bare macro: `apj` → "Astrophysical Journal"
    /// - Already full name: Returns unchanged
    ///
    /// - Parameter value: The journal field value from BibTeX
    /// - Returns: The expanded journal name, or the original if not a macro
    public static func expand(_ value: String) -> String {
        let trimmed = value.trimmingCharacters(in: .whitespaces)

        // Check for LaTeX macro format: \apj
        if trimmed.hasPrefix("\\") {
            let macroName = String(trimmed.dropFirst()).lowercased()
            if let fullName = macros[macroName] {
                return fullName
            }
        }

        // Check for bare macro name
        let lowercased = trimmed.lowercased()
        if let fullName = macros[lowercased] {
            return fullName
        }

        // Not a macro - return original
        return value
    }

    /// Checks if a string is a journal macro.
    public static func isMacro(_ value: String) -> Bool {
        let trimmed = value.trimmingCharacters(in: .whitespaces)

        // Check LaTeX format
        if trimmed.hasPrefix("\\") {
            let macroName = String(trimmed.dropFirst()).lowercased()
            return macros[macroName] != nil
        }

        // Check bare format
        return macros[trimmed.lowercased()] != nil
    }
}

File: crates/impress-bibtex/src/journal_macros.rs
//! Journal macro expansion
//!
//! Expands AASTeX journal macro abbreviations to full journal names.

use lazy_static::lazy_static;
use std::collections::HashMap;

lazy_static! {
    /// Dictionary mapping journal macros to full names.
    static ref MACROS: HashMap<&'static str, &'static str> = {
        let mut m = HashMap::new();

        // Major Astronomical Journals
        m.insert("aj", "Astronomical Journal");
        m.insert("apj", "Astrophysical Journal");
        m.insert("apjl", "Astrophysical Journal, Letters");
        m.insert("apjlett", "Astrophysical Journal, Letters");
        m.insert("apjs", "Astrophysical Journal, Supplement");
        m.insert("apjsupp", "Astrophysical Journal, Supplement");
        m.insert("mnras", "Monthly Notices of the Royal Astronomical Society");
        m.insert("aap", "Astronomy and Astrophysics");
        m.insert("astap", "Astronomy and Astrophysics");
        m.insert("aaps", "Astronomy and Astrophysics, Supplement");
        m.insert("pasp", "Publications of the Astronomical Society of the Pacific");
        m.insert("pasj", "Publications of the Astronomical Society of Japan");
        m.insert("pasa", "Publications of the Astronomical Society of Australia");

        // Review Journals
        m.insert("araa", "Annual Review of Astronomy and Astrophysics");
        m.insert("aapr", "Astronomy and Astrophysics Reviews");

        // Society Publications
        m.insert("baas", "Bulletin of the American Astronomical Society");
        m.insert("memras", "Memoirs of the Royal Astronomical Society");
        m.insert("qjras", "Quarterly Journal of the Royal Astronomical Society");
        m.insert("jrasc", "Journal of the Royal Astronomical Society of Canada");
        m.insert("memsai", "Memorie della Societa Astronomica Italiana");
        m.insert("iaucirc", "IAU Circulars");

        // Specialized Fields
        m.insert("icarus", "Icarus");
        m.insert("psj", "Planetary Science Journal");
        m.insert("solphys", "Solar Physics");
        m.insert("jcap", "Journal of Cosmology and Astroparticle Physics");
        m.insert("planss", "Planetary and Space Science");
        m.insert("grl", "Geophysical Research Letters");
        m.insert("jgr", "Journal of Geophysical Research");

        // Physics Journals
        m.insert("prl", "Physical Review Letters");
        m.insert("pra", "Physical Review A");
        m.insert("prb", "Physical Review B");
        m.insert("prc", "Physical Review C");
        m.insert("prd", "Physical Review D");
        m.insert("pre", "Physical Review E");
        m.insert("physrep", "Physics Reports");
        m.insert("physscr", "Physica Scripta");
        m.insert("nphysa", "Nuclear Physics A");

        // Optics and Spectroscopy
        m.insert("ao", "Applied Optics");
        m.insert("applopt", "Applied Optics");
        m.insert("jqsrt", "Journal of Quantitative Spectroscopy and Radiative Transfer");
        m.insert("procspie", "Proceedings of the SPIE");

        // General Science
        m.insert("nat", "Nature");
        m.insert("sci", "Science");
        m.insert("ssr", "Space Science Reviews");
        m.insert("apss", "Astrophysics and Space Science");

        // Regional Journals
        m.insert("actaa", "Acta Astronomica");
        m.insert("azh", "Astronomicheskii Zhurnal");
        m.insert("sovast", "Soviet Astronomy");
        m.insert("caa", "Chinese Astronomy and Astrophysics");
        m.insert("cjaa", "Chinese Journal of Astronomy and Astrophysics");
        m.insert("rmxaa", "Revista Mexicana de Astronomia y Astrofisica");
        m.insert("bac", "Bulletin of the Astronomical Institutes of Czechoslovakia");
        m.insert("bain", "Bulletin Astronomical Institute of the Netherlands");
        m.insert("zap", "Zeitschrift fuer Astrophysik");

        // Other
        m.insert("na", "New Astronomy");
        m.insert("nar", "New Astronomy Review");
        m.insert("skytel", "Sky & Telescope");
        m.insert("aplett", "Astrophysics Letters");
        m.insert("apspr", "Astrophysics Space Physics Research");
        m.insert("fcp", "Fundamental Cosmic Physics");
        m.insert("gca", "Geochimica et Cosmochimica Acta");
        m.insert("jcp", "Journal of Chemical Physics");

        m
    };
}

/// Expand a journal macro to its full name
pub fn expand_journal_macro(value: String) -> String {
    let trimmed = value.trim();

    // Check for LaTeX macro format: \apj
    if let Some(stripped) = trimmed.strip_prefix('\\') {
        let macro_name = stripped.to_lowercase();
        if let Some(&full_name) = MACROS.get(macro_name.as_str()) {
            return full_name.to_string();
        }
    }

    // Check for bare macro name
    let lowercased = trimmed.to_lowercase();
    if let Some(&full_name) = MACROS.get(lowercased.as_str()) {
        return full_name.to_string();
    }

    // Not a macro - return original
    value
}

#[cfg(feature = "uniffi")]
#[uniffi::export]
pub fn expand_journal_macro_ffi(value: String) -> String {
    expand_journal_macro(value)
}

/// Check if a value is a journal macro
pub fn is_journal_macro(value: String) -> bool {
    let trimmed = value.trim();

    // Check LaTeX format
    if let Some(stripped) = trimmed.strip_prefix('\\') {
        let macro_name = stripped.to_lowercase();
        return MACROS.contains_key(macro_name.as_str());
    }

    // Check bare format
    MACROS.contains_key(trimmed.to_lowercase().as_str())
}

#[cfg(feature = "uniffi")]
#[uniffi::export]
pub fn is_journal_macro_ffi(value: String) -> bool {
    is_journal_macro(value)
}

/// Get all journal macro names
pub fn get_all_journal_macro_names() -> Vec<String> {
    MACROS.keys().map(|k| k.to_string()).collect()
}

#[cfg(feature = "uniffi")]
#[uniffi::export]
pub fn get_all_journal_macro_names_ffi() -> Vec<String> {
    get_all_journal_macro_names()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_expand_latex_format() {
        assert_eq!(
            expand_journal_macro("\\apj".to_string()),
            "Astrophysical Journal"
        );
        assert_eq!(
            expand_journal_macro("\\mnras".to_string()),
            "Monthly Notices of the Royal Astronomical Society"
        );
    }

    #[test]
    fn test_expand_bare_format() {
        assert_eq!(
            expand_journal_macro("apj".to_string()),
            "Astrophysical Journal"
        );
    }

    #[test]
    fn test_expand_case_insensitive() {
        assert_eq!(
            expand_journal_macro("APJ".to_string()),
            "Astrophysical Journal"
        );
    }

    #[test]
    fn test_expand_unknown_returns_original() {
        assert_eq!(
            expand_journal_macro("Nature Physics".to_string()),
            "Nature Physics"
        );
    }

    #[test]
    fn test_is_macro() {
        assert!(is_journal_macro("apj".to_string()));
        assert!(is_journal_macro("\\apj".to_string()));
        assert!(!is_journal_macro("Nature".to_string()));
    }

    #[test]
    fn test_get_all_macro_names() {
        let names = get_all_journal_macro_names();
        assert!(names.len() > 50);
        assert!(names.contains(&"apj".to_string()));
    }
}
